daemon/containerd/image_delete.go

```diff
@@ -58,12 +61,26 @@ func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, force,
 		return nil, err
 	}
 
+	possiblyDeletedConfigs := map[digest.Digest]struct{}{}
+	if err := i.walkPresentChildren(ctx, desc, func(_ context.Context, d ocispec.Descriptor) {
+		if images.IsConfigType(d.MediaType) {
+			possiblyDeletedConfigs[d.Digest] = struct{}{}
+		}
+	}); err != nil {
+		return nil, err
+	}
+
 	imgID := string(desc.Digest)
 	err = i.client.ImageService().Delete(ctx, ref.String(), images.SynchronousDelete())
 	if err != nil {
 		return nil, err
 	}
 
+	// Workaround for: https://github.com/moby/buildkit/issues/3797
+	if err := i.unleaseSnapshotsFromDeletedConfigs(ctx, possiblyDeletedConfigs); err != nil {
+		logrus.WithError(err).Warn("failed to unlease snapshots")
```

Comment:
What are the consequences of failing to unlease the snapshots? I get the impression that a rootfs snapshot could leak. And since the image has already been deleted at this point, the user would not be able to retry. I presume that the only recourse would be to use a low-level tool such as `ctr` to poke the containerd API directly?

Would it be possible to unlease the snapshots before the image is deleted such that the image is only deleted if the unleasing succeeds?

Reply:
It will "leak" in a sense that the lease attached by the history entry will still be there, so as long as the history entry is not deleted.
So eventually the snapshot will get deleted after build history entry is removed (50 entries is the maximum by default).
Also, the failure (except context cancellation) shouldn't happen unless there's something really wrong with the content store state, which probably means that the user will have to clear it anyway.

Removing the lease before deleting images isn't good either, because the operation can be cancelled which means that we would unlease content which actually didn't get deleted.

I think the approach suggested by @tonistiigi suggested would be the best here.